compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -2560,37 +2566,28 @@ TR::Register *OMR::X86::TreeEvaluator::OverflowCHKEvaluator(TR::Node *node, TR::
       }
 
    cg->setVMThreadRequired(true);
-   //the BBStartEvaluator will generate the label but in this case the catch block might not been evaluated yet
-   TR::Node * bbstartNode = overflowCatchBlock->getEntry()->getNode();
-   TR_ASSERT((bbstartNode->getOpCodeValue() == TR::BBStart), "catch block entry %p must be TR::BBStart\n", bbstartNode);
-   if (!bbstartNode->getLabel()) 
-      {
-      TR::LabelSymbol *label = generateLabelSymbol(cg);
-      bbstartNode->setLabel(label);
-      }
+   }
 
-   bool isUnsigned = false;
-   switch (node->getOverflowCheckOperation())
+TR::Register *OMR::X86::TreeEvaluator::overflowCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-       case TR::buadd:
-       case TR::cadd:
-       case TR::iuadd:
-       case TR::luadd:
-       case TR::busub:
-       case TR::csub:
-       case TR::iusub:
-       case TR::lusub:
-           isUnsigned = true;
-           break;
+   TR::Block *overflowCatchBlock = TR::TreeEvaluator::getOverflowCatchBlock(node, cg);
+   TR::TreeEvaluator::genArithmeticInstructionsForOverflowCHK(node, cg);
+   generateLabelInstruction(JO4, node, overflowCatchBlock->getEntry()->getNode()->getLabel(), cg);
+   cg->setVMThreadRequired(false);
+   cg->decReferenceCount(node->getFirstChild());
+   return NULL;
    }
 
-   TR_X86OpCodes opcode = isUnsigned? JB4: JO4; 
-   generateLabelInstruction(opcode, node, overflowCatchBlock->getEntry()->getNode()->getLabel(), cg);
+TR::Register *OMR::X86::TreeEvaluator::unsignedOverflowCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
```

Comment:
This code looks virtually identical to overflowCHKEvaluator. Could you use the same function and just pick JO4 versus JB4 based on the node's opcode?